tests/integration_tests/test_wandb.py

```diff
@@ -28,114 +31,268 @@ def _multiobjective_func(trial: optuna.trial.Trial) -> Tuple[float, float]:
-    assert wandb.log.call_count == target_n_trials
+    assert wandb.run.log.call_count == target_n_trials
+
+    wandbc = WeightsAndBiasesCallback(as_multirun=True)
+    wandb.run.reset_mock()
+    _wrapped_logging_func = wandbc.track_in_wandb()(
+        functools.partial(logging_objective_func, log_func=wandb.run.log)
+    )
+
+    study.optimize(_wrapped_logging_func, n_trials=target_n_trials, callbacks=[wandbc])
+
+    assert wandb.run.log.call_count == 2 * target_n_trials
+
+    wandb.run = None
+    study.optimize(_objective_func, n_trials=target_n_trials, callbacks=[wandbc])
+    assert wandb.init().log.call_count == target_n_trials
 
 
 @pytest.mark.parametrize(
-    "metric,expected", [("value", ["x", "y", "value"]), ("foo", ["x", "y", "foo"])]
+    "metric,as_multirun,expected",
+    [("value", False, ["x", "y", "value"]), ("foo", True, ["x", "y", "foo", "trial_number"])],
 )
 @mock.patch("optuna.integration.wandb.wandb")
-def test_values_registered_on_epoch(wandb: mock.Mock, metric: str, expected: List[str]) -> None:
+def test_values_registered_on_epoch(
+    wandb: mock.MagicMock, metric: str, as_multirun: bool, expected: List[str]
+) -> None:
+    def assert_call_args(log_func: mock.MagicMock, regular: bool) -> None:
+        call_args = log_func.call_args
+        assert list(call_args[0][0].keys()) == expected
+        assert call_args[1] == {"step": 0 if regular else None}
 
-    wandb.log = mock.MagicMock()
+    wandb.sdk.wandb_run.Run = mock.MagicMock
+
+    if as_multirun:
+        wandb.run = None
+        log_func = wandb.init().log
+    else:
+        log_func = wandb.run.log
 
-    wandbc = WeightsAndBiasesCallback(metric_name=metric)
     study = optuna.create_study()
+    wandbc = WeightsAndBiasesCallback(metric_name=metric, as_multirun=as_multirun)
     study.optimize(_objective_func, n_trials=1, callbacks=[wandbc])
+    assert_call_args(log_func, bool(wandb.run))
 
-    kall = wandb.log.call_args
-    assert list(kall[0][0].keys()) == expected
-    assert kall[1] == {"step": 0}
+
+@pytest.mark.parametrize("metric,expected", [("foo", ["x", "y", "foo", "trial_number"])])
+@mock.patch("optuna.integration.wandb.wandb")
+def test_values_registered_on_epoch_with_logging(
+    wandb: mock.MagicMock, metric: str, expected: List[str]
+) -> None:
+
+    wandb.sdk.wandb_run.Run = mock.MagicMock
+
+    study = optuna.create_study()
+    wandbc = WeightsAndBiasesCallback(metric_name=metric, as_multirun=True)
+    _wrapped_func = wandbc.track_in_wandb()(
+        functools.partial(logging_objective_func, log_func=wandb.run.log)
+    )
+
+    study.enqueue_trial({"x": 2, "y": 25})
+    study.optimize(_wrapped_func, n_trials=1, callbacks=[wandbc])
+
+    logged_metric = wandb.run.log.mock_calls[0][1][0]
+
+    call_args = wandb.run.log.call_args
+    assert list(call_args[0][0].keys()) == expected
+    assert call_args[1] == {"step": 0}
+    assert logged_metric == {"result": 0}
```

Comment:
I think that we can be a bit clearer what is being tested here - `wandb.run.log` should be called twice (from decorator and from callback), so IMO we can test following:
* Number of calls
* Payload of the first call
* Payload of the second call

```suggestion
    logged_in_decorator = wandb.run.log.mock_calls[0][1][0]
    logged_in_callback = wandb.run.log.mock_calls[1][1][0]
    assert len(wandb.run.log.mock_calls) == 2
    assert list(logged_in_decorator) == ["result"]
    assert list(logged_in_callback) == expected

    call_args = wandb.run.log.call_args
    assert call_args[1] == {"step": 0}
```

Also tesing for metric + values would be nicer, but I guess we would need to remove parametrization. What do you think?

Reply:
Can you please elaborate on what you mean by testing metric + values? An example would be very helpful.

Reply:
Not sure what you are saying, can you elaborate it a bit?

Reply:
At the current stage test asserts if metric names are correct:
```python
assert list(logged_in_decorator) == ["result"]
assert list(logged_in_callback) == expected
```
I think it's because we are passing `expected` as a list of metric names via `mark.parametrize` decorator. IMO we could pass a dict of metric names and respective values and assert something like following:
```python
assert logged_in_decorator == {"result": 0}
assert logged_in_callback == {"x": 1.0, "y": 0.0, "foo":...}
```
But I'm fine with leaving this as future work.